gflownet.py

```diff
@@ -826,23 +832,52 @@ def trajectorybalance_loss(self, it, batch):
             Loss of the intermediate nodes only
         """
         # Unpack batch
-        _, _, rewards, parents, actions, done, path_id_parents, _ = zip(*batch)
+        (
+            state,
+            last_action,
+            reward,
+            parents,
+            parents_a,
+            done,
+            path_id_parents,
+            _,
+        ) = zip(*batch)
+        indices = list(map(index_where_equal, zip(parents_a, last_action)))
+        indices = torch.cat(indices)
+        parent_of_state_in_traj = list(map(select_index, zip(parents, indices)))
+
         path_id = torch.cat([el[:1] for el in path_id_parents])
-        rewards, parents, actions, done, path_id_parents = map(
-            torch.cat, [rewards, parents, actions, done, path_id_parents]
+
+        state, reward, parents, last_action, done = map(
+            torch.cat,
+            [
+                state,
+                reward,
+                parent_of_state_in_traj,
+                last_action,
+                done,
+            ],
         )
-        # Log probs of each (s, a)
-        logprobs = self.logsoftmax(self.model(parents))[
-            torch.arange(parents.shape[0]), actions
-        ]
+        # Forward and Backward log probs of each (s, a)
+        logprobs_f_1 = self.logsoftmax(
+            self.model(parents)[..., : len(self.env.action_space) + 1]
+        )
+        logprobs_f = logprobs_f_1[torch.arange(parents.shape[0]), last_action]
+        logprobs_b_1 = self.logsoftmax(
+            self.model(state)[..., len(self.env.action_space) + 1 :]
+        )
+        logprobs_b = logprobs_b_1[torch.arange(state.shape[0]), last_action]
         # Sum of log probs
-        sumlogprobs = tf(
+        sumlogprobs_f = tf(
+            torch.zeros(len(torch.unique(path_id, sorted=True)))
+        ).index_add_(0, path_id, logprobs_f)
+        sumlogprobs_b = tf(
             torch.zeros(len(torch.unique(path_id, sorted=True)))
-        ).index_add_(0, path_id_parents, logprobs)
+        ).index_add_(0, path_id, logprobs_b)
         # Sort rewards of done states by ascending path id
-        rewards = rewards[done.eq(1)][torch.argsort(path_id[done.eq(1)])]
-        # Trajectory balance loss
-        loss = (self.Z.sum() + sumlogprobs - torch.log((rewards))).pow(2).mean()
+        reward = reward[done.eq(1)][torch.argsort(path_id[done.eq(1)])]
+        # Trajectory balance loss sumlogprobs_b
+        loss = (self.Z.sum() + sumlogprobs_f - torch.log(reward)).pow(2).mean()
```

Comment:
`sumlogprobs_b` is never added (actually subtracted) to the final loss!

Reply:
Oh yeah, sorry I was testing a bunch of stuff because the loss wasn't converging as expected. I had removed that for one of the experiments. Forgot to add it back